### PR TITLE
Support AWT image resize via new AWT extension

### LINKED ISSUE #19789: AWT Graphics does not seem to work in native image
### Describe the bug

So basically I'm trying to resize png picture in quarkus app and I'm aware that there were several problems with AWT library. I tried following [this](https://github.com/graalvm/mandrel/issues/180#issuecomment-734873511) comment but so far I cannot make this thing to work. The main culprit in stack trace seems to be this
```
Caused by: java.lang.UnsupportedOperationException: Not implemented yet for GraalVM native images
        at java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment(GraphicsEnvironment.java:15)
        at java.awt.image.BufferedImage.createGraphics(BufferedImage.java:1181)
        at java.awt.image.BufferedImage.getGraphics(BufferedImage.java:1170)
        at com.test.ImageResource.resizeImage(ImageResource.java:63)
        at com.test.ImageResource.save(ImageResource.java:25)
```

I did create config files using agent like mentioned in that thread (you can check them [here](https://github.com/habimt/rest-graphics/tree/master/src/main/resources/META-INF/native-image)) but I'm wondering if perhaps there's something wrong with them.

Config files were crated the following way:
```
$GRAALVM_HOME/bin/java -agentlib:native-image-agent=config-output-dir=native-image-config -Djava.awt.headless=true -jar target/quarkus-app/quarkus-run.jar
```

### How to Reproduce?
clone this https://github.com/habimt/rest-graphics and run `./mvnw verify -Pnative`

### Output of `java -version`

OpenJDK Runtime Environment (build 11.0.12+7)

### GraalVM version (if different from Java)

GraalVM 21.2.0 Java 11 CE (Java Version 11.0.12+6-jvmci-21.2-b08), also tried on 21.2.0.0-Final Mandrel Distribution (Java Version 11.0.12+7)

### Quarkus version or git rev
2.1.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Apache Maven 3.8.1

### PR DESCRIPTION
**Pull request by galderz**
This PR replaces #20148.

It enables AWT use cases, such as image resizing, using a dedicated AWT extension.

Note: the extension was generated with the `create-extension` command of the quarkus maven plugin.

### CODE REVIEW

**Comment by gastaldi** [comment]
@galderz can you rebase with the latest `main`? There are some conflicts

**Comment by gastaldi** [comment]
Also it would be great if you could also squash your commits

**Comment by galderz** [comment]
Yeah yeah, on it already...

**Comment by galderz** [comment]
@gastaldi Done. Let's see what CI says

File: extensions/awt/runtime/pom.xml
Comment (gastaldi) [inline extensions/awt/runtime/pom.xml:12]: Add a `<description>` here, it will be used in the resulting quarkus-extension.yaml and our tooling

File: extensions/awt/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (geoand) [reply extensions/awt/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: Maybe it should be `experimental`?

File: integration-tests/awt/src/test/java/io/quarkus/awt/it/AwtImageResizeTest.java
Comment (geoand) [inline integration-tests/awt/src/test/java/io/quarkus/awt/it/AwtImageResizeTest.java:8]: Please avoid star imports

File: extensions/awt/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gastaldi) [reply extensions/awt/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: Ah yes, good idea

File: extensions/awt/runtime/src/main/java/io/quarkus/awt/runtime/graal/AwtFeature.java
Comment (Sanne) [inline extensions/awt/runtime/src/main/java/io/quarkus/awt/runtime/graal/AwtFeature.java:10]: :+1:

**Comment by galderz** [comment]
@gastaldi @geoand All comments addressed.

**Comment by Sanne** [comment]
I have a general question.. how do you expect users to know they need to add this extension?

Should we consider it a "technical extension" and have other extensions such as those doing image manipulations depend on it?

Or perhaps it belongs in all applications, and we should pull it in automatically for all users?

I'm inclined to think it shoud be pulled in for every user automatically, provided there is no overhead. I'd normally expect this to not have any overhead, but you seem to have registered some classes for reflective access so that might force some code inclusion which would otherwise be identified as dead code. Why the reflective registrations?

**Comment by geoand** [comment]
The test also needs to be added to the native test configuration. I don't remember exactly where it is (and I'm on a phone right now), but @gastaldi can give you the exact location

**Comment by gastaldi** [comment]
> The test also needs to be added to the native test configuration. I don't remember exactly where it is (and I'm on a phone right now), but @gastaldi can give you the exact location

It's in https://github.com/quarkusio/quarkus/blob/main/.github/native-tests.json

**Comment by galderz** [comment]
> I have a general question.. how do you expect users to know they need to add this extension?
> 
> Should we consider it a "technical extension" and have other extensions such as those doing image manipulations depend on it?
> 
> Or perhaps it belongs in all applications, and we should pull it in automatically for all users?

Is there such big proportion of Quarkus apps doing image manipulations? Unless there is, I'd keep this as something that people bring in when they need it.

Or are you seeing this as something that should work out of the box?

> I'm inclined to think it shoud be pulled in for every user automatically, provided there is no overhead. I'd normally expect this to not have any overhead, but you seem to have registered some classes for reflective access so that might force some code inclusion which would otherwise be identified as dead code. Why the reflective registrations?

The reflection registration for `sun.awt.X11.XToolkit` are for the [`Toolkit.getDefaultToolkit()`](https://github.com/openjdk/jdk11u-dev/blob/master/src/java.desktop/share/classes/java/awt/Toolkit.java#L580-L617) code. IIRC `sun.awt.X11GraphicsEnvironment` is for the [`GraphicsEnvironment`](https://github.com/openjdk/jdk11u-dev/blob/master/src/java.desktop/share/classes/java/awt/GraphicsEnvironment.java#L91-L122) initialization.

I would expect the JNI registrations to have some cost too, and there are more of them compared to reflection registrations. My gut feeling is that you'll need even more JNI registrations for other AWT use cases.

**Comment by galderz** [comment]
Actually @sanne, I think might have a point:

The hibernate-orm-panache native image build didn't complete, see [extract maven logs](https://gist.github.com/galderz/5c223207b69e1754472bea2ca6c59f26):

```
"ForkJoinPool-2-worker-3" Id=20 in RUNNABLE
    at java.desktop@11.0.12/sun.java2d.loops.FillRect.<clinit>(FillRect.java:79)
```

What's happening is that as a result of removing [`Java2DSubstitutions`](https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/graal/Java2DSubstitutions.java), a load of JAXB graphics related code makes it to the universe, e.g.

```
│   │       │   │       │       │           │   │   │   │       │           │   │   │   │   │   │   │   │   │   │           │   ├── directly calls com.sun.xml.bind.v2.model.impl.RuntimeBuiltinLeafInfoImpl$10$1.<init>(com.sun.xml.bind.v2.model.impl.RuntimeBuiltinLeafInfoImpl$10):void id=103127 @bci=21 
│   │       │   │       │       │           │   │   │   │       │           │   │   │   │   │   │   │   │   │   │           │   │   └── directly calls java.awt.Component.<init>():void id=104120 @bci=6 
```

To solve the problem above, we'd need an extension that is applied to all that takes care of the runtime initialization of all those packages. This could also be done directly in core, by moving `io.quarkus.awt.runtime.graal.AwtFeature` to core.

You could still have an optional extension for those who really want to use awt with reflection+jni registrations (or any optimised versions of those).

Thoughts?

**Comment by Sanne** [comment]
@galderz yes this is a bit concerning. What if we move the build items into quarkus - core.. is the overhead of these JNI and reflection registrations high?

Ideally - but I have no idea how hard and how much work that would be - these should all be made conditional to certain symbols being reacheable, and perhaps the patch belongs in GraalVM rather than in Quarkus as anyone could use AWT (but we can certainly have a preview here).

**Comment by galderz** [comment]
> @galderz yes this is a bit concerning. What if we move the build items into quarkus - core.. is the overhead of these JNI and reflection registrations high?

After some thought, I've gone for a split approach: runtime init packages are defined in core, then all reflection/JNI registrations happen in `extensions/awt`. This would solve the problem. I've explained the details for this split in the javadoc of `AwtFeature`:

```
 * Technically, this should live in extensions/awt,
 * but currently all code that relies on JAXB
 * requires at the very least sun.java2d package to be runtime initialized.
 *
 * Having sun.java2d code initialized at build time caused issues,
 * which is why a substitution was set in place to avoid such code making it to the binary:
 * https://github.com/quarkusio/quarkus/commit/ef87e5567cf3ac462a3f12aad4b5b530d9220223
 *
 * So, as long as JAXB graphics code has not been excluded completely from JAXB,
 * it is safer to define all image related packages to be runtime initialized directly in core.
```

> Ideally - but I have no idea how hard and how much work that would be - these should all be made conditional to certain symbols being reacheable, and perhaps the patch belongs in GraalVM rather than in Quarkus as anyone could use AWT (but we can certainly have a preview here).

+1. Let's nail things first in Quarkus and then can propose them to GraalVM.

**Comment by galderz** [comment]
Pushed a couple of commits to fix the tests that also failed in previous PR. They've fixed in different way this time:
* `ImageIOTestCase` has been moved to integration tests for awt.
* Tika extension now depends on awt extension. It also adds a feature to mark which packages need to be runtime initialized.

The only thing remaining then would be documentation. Where would it go?

**Comment by geoand** [comment]
When everything is done, please squash the formatting commits

**Comment by galderz** [comment]
> When everything is done, please squash the formatting commits

Yeah sure. Just waiting for a final CI run to make sure I've not forgotten anything.

**Comment by galderz** [comment]
Squashed and pushed. This is ready to go in now.

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by geoand** [comment]
@Sanne I'll leave the final decision up to you since you've been following this closely

**Review by Sanne — APPROVED** [review]

**Comment by Sanne** [comment]
> @Sanne I'll leave the final decision up to you since you've been following this closely

damn :)

thanks all! merged it

**Comment by geoand** [comment]
One more question: Should we backport this into `2.3`?

**Comment by galderz** [comment]
@geoand Yeah, the impact on existing code is minimal (just add awt/imageio/java2d packages for runtime init). For the rest of the code to kick in, an extension has to depend on `awt` specifically.

**Comment by geoand** [comment]
I am removing the backport label as a concequence of https://github.com/quarkusio/quarkus/issues/11563

**Comment by galderz** [comment]
@geoand Can someone create an issue for #11563 and assign it to me?  I've been digging into https://github.com/graalvm/mandrel/issues/292

**Comment by geoand** [comment]
Done: https://github.com/quarkusio/quarkus/issues/20491

**Pull request merged**
